### ISSUE TITLE
Add additional attributes to ManifestBuild descriptors for publishing

### ISSUE THREAD

**Issue by ysndr** [flox/flox#2237]
**Describe the feature:**

Extend the build descriptor with the following fields

```
build.<name> = {
  version = null | STRING
  description = null | STRING
  license = null | STRING
}
```

* `version`: packages should _always_ have a versioned pname, meaning they build to a store path `/nix/store/<hash>-<name>-<version>` for clarity purposes. (Luckily we dont need to do the opposite operation of splitting pname into name and version with these). Since the manifest builder is reading package information from the manifest directly, we may need to do some refacoring to ensure a single source of thruth or align the default in both the publishing logic (et al, in rust) and the make based builder.
* `description`: textual description of what the built package does, used on Floxhub and search
* license: a license string as we currently use for nixpkgs packages.

**Comment by mkenigs** [comment]
>     * `version`: packages should _always_ have a versioned pname
What do we default to?

**Comment by limeytexan** [comment]
> What do we default to?

IIRC last week we discussed setting it to some obviously "unset" string, like `unknown` or `noversion`. Whatever it is it shouldn't be possible to interpret it as a valid semantic version (i.e. we shouldn't use `0.0.0`).

**Comment by ysndr** [comment]
> `version = null | STRING`

we've seen a desire to support "dynamic" versions, e.g. through a command.

Question is how we can represent that.


```
version := null | STRING | Dynamic

Dynamic := {
  command := STRING
}
```

**Comment by billlevine** [comment]
In thinking about this w.r.t. publish, another way to think about it is to add them for build, and to include them in the build process such that `nix derivation show` presents the information set.  Assuming that is possible and desirable, publish can then use that to get the info needed for publish ensuring consistency.

**Comment by ysndr** [comment]
Before we conflate two things here:
This issue asks to add attributes to the manifest only, probably despite it likely being useful, without dynamic versions yet.
This will be followed up (pending discussion on eng design) with work in 
* the package builder, to read out the version and provide it to the build
* the cli to integrate the given information during publish

Come _dynamic versions_, those will likely be "generated" within the package builder and need to be passed "up and down" the stack into the actual build, as well as the cli for publish. We might abuse the derivation format for this or add a separate output/package that communicates said information, pending -again- eng discussion with @limeytexan @billlevine and team CLI / probably me.

**Comment by limeytexan** [comment]
Thanks @ysndr - agree the expected place for invoking a command which determines the version number is within the build itself, so we need to have a way for the builder to pass back this rich information to the CLI after a build.

At the moment we have the following:

1. the output of `nix build --json` which includes an array of items each containing:
    * the path to the derivation
    * the path to all outputs
1. `result-*` symlinks pointing to all the outputs (which can be considered a dup of the previous item)

At present we are not passing back this JSON data to the CLI, but we could (probably by way of a named pipe created prior to the build), and we can also augment that data with other data such as the `pname` and `version` used for the build prior to passing it back to the CLI.

By way of example, here's a sample output of `nix build --json`:
```
[
  {
    "drvPath": "/nix/store/7apss80hr6h3w9w3hf9bxpabk18z0vp4-foo-0.0.0.drv",
    "outputs": {
      "out": "/nix/store/hd7dxi7kw8caq16088pc8ljq7ygp0bww-foo-0.0.0"
    }
  }
]
```

... and I'd think that we could update this to instead return:
```
[
  {
    "drvPath": "/nix/store/7apss80hr6h3w9w3hf9bxpabk18z0vp4-foo-0.0.0.drv",
    "outputs": {
      "out": "/nix/store/hd7dxi7kw8caq16088pc8ljq7ygp0bww-foo-0.0.0"
    },
    "pname": "foo",
    "version": "0.0.0"
  }
]
```

... or something like that.

**Comment by limeytexan** [comment]
The other thing to note is that Yannik was right with his [comment](https://github.com/flox/flox/issues/2237#issuecomment-2447364518) above - the version string starts in the manifest, but can be either a string or a command. The build consumes this data by way of the `manifest.lock`, and then publish should honor the version string that comes back from the build (which can be the same or different).

**Comment by zmitchell** [comment]
Unclear from the current discussion what we want to support _today_. Is the `version` field in the manifest a static string, or do we also allow it to be a script? This changes how we document the feature.

**Comment by billlevine** [comment]
> Unclear from the current discussion what we want to support _today_. Is the `version` field in the manifest a static string, or do we also allow it to be a script? This changes how we document the feature.

From the publish meeting today we said it can be a static string for now.   We will want the dynamic evaluation later but that doesn't have to be now and brings with it some complexity w.r.t. the build part.

**Comment by billlevine** [comment]
> The other thing is that sometimes projects will have multiple licenses e.g. MIT and APACHE. The ticket lists a single string, but it probably makes sense to be a list instead.

Agreed, where possible we should match the nix model.  I do think `license` should be a list of strings.
